### PR TITLE
Adding transaction batch support in cosmos encryption

### PR DESCRIPTION
**Pull request by simplynaveen20**
This PR will add support for executing transaction batch in async and sync encryption container.

API name ,  its parameter (CosmosBatch) and response(CosmosBatchResponse) will remain the same as of core sdk.
We are doing encryption once cx call executeCosmosBatch by creating new list of ItemBatchOperations, and before returning  CosmosBatchResponse decrypting all existing CosmosBatchOperationResult.

`  public Mono<CosmosBatchResponse> executeCosmosBatch(CosmosBatch cosmosBatch, CosmosBatchRequestOptions requestOptions)`

` public Mono<CosmosBatchResponse> executeCosmosBatch(CosmosBatch cosmosBatch) `

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 1]
```diff
@@ -243,6 +243,10 @@ public EncryptionSettings getEncryptionSettings() {
     }
 
     public Mono<byte[]> encrypt(ObjectNode itemJObj) {
+        return encryptObjectNode(itemJObj).flatMap(encryptedObjectNode -> Mono.just(EncryptionUtils.serializeJsonToByteArray(EncryptionUtils.getSimpleObjectMapper(), encryptedObjectNode)));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:246]: you don't need flatMap here. 

Here you are wrapping the result with `Mono.just` then you flatten it with `flatMap` you don't need `flatMap `nor the `Mono.just `here

this is like doing `sqrt(x^2)` just do `x` :-)

```suggestion
        return encryptObjectNode(itemJObj).map(encryptedObjectNode -> EncryptionUtils.serializeJsonToByteArray(EncryptionUtils.getSimpleObjectMapper(), encryptedObjectNode));
```

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 2]
```diff
@@ -270,7 +274,7 @@ public EncryptionSettings getEncryptionSettings() {
                 }
             }
             Mono<List<Void>> listMono = Flux.mergeSequential(encryptionMonoList).collectList();
-            return listMono.flatMap(ignoreVoid -> Mono.just(EncryptionUtils.serializeJsonToByteArray(EncryptionUtils.getSimpleObjectMapper(), itemJObj)));
+            return listMono.flatMap(ignoreVoid -> Mono.just(itemJObj));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:277]: ditto on `flatMap` and `Mono.just`

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 3]
```diff
@@ -357,6 +361,10 @@ public void encryptAndSerializeProperty(EncryptionSettings encryptionSettings, O
     }
 
     public Mono<byte[]> decrypt(ObjectNode itemJObj) {
+        return decryptObjectNode(itemJObj).flatMap(decryptedObjectNode -> Mono.just(EncryptionUtils.serializeJsonToByteArray(EncryptionUtils.getSimpleObjectMapper(), decryptedObjectNode)));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:364]: ditto on `flatMap` and `Mono.just`

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 4]
```diff
@@ -387,7 +395,7 @@ public void encryptAndSerializeProperty(EncryptionSettings encryptionSettings, O
                 }
             }
             Mono<List<Void>> listMono = Flux.mergeSequential(encryptionMonoList).collectList();
-            return listMono.flatMap(aVoid -> Mono.just(EncryptionUtils.serializeJsonToByteArray(EncryptionUtils.getSimpleObjectMapper(), itemJObj)));
+            return listMono.flatMap(aVoid -> Mono.just(itemJObj));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:398]: ditto

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java [hunk 5]
```diff
@@ -594,6 +607,182 @@ private void setRequestHeaders(CosmosChangeFeedRequestOptions requestOptions) {
...
+                    encryptionProcessor.encryptObjectNode(objectNode).map(encryptedItem -> {
+                        return new ItemBatchOperation(
+                            itemBatchOperation.getOperationType(),
+                            itemBatchOperation.getId(),
+                            itemBatchOperation.getPartitionKeyValue(),
+                            itemBatchOperation.getRequestOptions(),
+                            encryptedItem
+                        );
+                    });
+            } else {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java:698]: trying to understand the `else` branch.

in which scenario we go to the `else` branch? `itemBatchOperation.getItem() != null` which scenario is this? is this for `delete` and `read` only?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java [hunk 6]
```diff
@@ -594,6 +607,182 @@ private void setRequestHeaders(CosmosChangeFeedRequestOptions requestOptions) {
...
+                if (objectNode != null) {
+                    decryptMonoList.add(encryptionProcessor.decryptObjectNode(objectNode).flatMap(jsonNode -> {
+                        this.cosmosBatchOperationResultAccessor.setResourceObject(cosmosBatchOperationResult, jsonNode);
+                        return Mono.empty();
+                    }));
+                }
+            }
+
+            Mono<List<Void>> listMono = Flux.mergeSequential(decryptMonoList).collectList();
+            return listMono.flatMap(aVoid -> Mono.just(cosmosBatchResponse));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java:752]: is this guaranteed that `listMono` is never `Mono.empty`?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 1 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:246]: Thanks for noticing, it was copy paste during refactoring from previous code, moved to map now

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 2 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:277]: done

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 4 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:398]: done

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 3 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:364]: done

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java (hunk 5 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java:698]: Yes read and delete do not have the item object in itemBatchOperation

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java (hunk 6 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java]: Yes listMono is never Mono.empty. Tested it by single batch with delete operation , also by  manually passing  List<Mono<Void>> decryptMonoList = new ArrayList<>() directly

**Review by moderakh — APPROVED** [review]
Thanks Naveen

**Pull request merged**